[SPARK-23623] [SS] Avoid concurrent use of cached consumers in CachedKafkaConsumer

tdas opened the pull request:
## What changes were proposed in this pull request?

CacheKafkaConsumer in the project `kafka-0-10-sql` is designed to maintain a pool of KafkaConsumers that can be reused. However, it was built with the assumption there will be only one task using trying to read the same Kafka TopicPartition at the same time. Hence, the cache was keyed by the TopicPartition a consumer is supposed to read. And any cases where this assumption may not be true, we have SparkPlan flag to disable the use of a cache. So it was up to the planner to correctly identify when it was not safe to use the cache and set the flag accordingly. 

Fundamentally, this is the wrong way to approach the problem. It is HARD for a high-level planner to reason about the low-level execution model, whether there will be multiple tasks in the same query trying to read the same partition. Case in point, 2.3.0 introduced stream-stream joins, and you can build a streaming self-join query on Kafka. It's pretty non-trivial to figure out how this leads to two tasks reading the same partition twice, possibly concurrently. And due to the non-triviality, it is hard to figure this out in the planner and set the flag to avoid the cache / consumer pool. And this can inadvertently lead to ConcurrentModificationException ,or worse, silent reading of incorrect data.

Here is a better way to design this. The planner shouldnt have to understand these low-level optimizations. Rather the consumer pool should be smart enough avoid concurrent use of a cached consumer. Currently, it tries to do so but incorrectly (the flag inuse is not checked when returning a cached consumer, see [this](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala#L403)). If there is another request for the same partition as a currently in-use consumer, the pool should automatically return a fresh consumer that should be closed when the task is done. Then the planner does not have to have a flag to avoid reuses.

This PR is a step towards that goal. It does the following. 
- There are effectively two kinds of consumer that may be generated 
  - Cached consumer - this should be returned to the pool at task end
  - Non-cached consumer - this should be closed at task end
- A trait called KafkaConsumer is introduced to hide this difference from the users of the consumer so that the client code does not have to reason about whether to stop and release. They simply called `val consumer = KafkaConsumer.acquire` and then `consumer.release()`.
- If there is request for a consumer that is in-use, then a new consumer is generated.
- If there is a concurrent attempt of the same task, then a new consumer is generated, and the existing cached consumer is marked for close upon release. 
- In addition, I renamed the classes because CachedKafkaConsumer is a misnomer given that what it returns may or may not be cached. 

This PR does not remove the planner flag to avoid reuse to make this patch safe enough for merging in branch-2.3. This can be done later in master-only.



## How was this patch tested?
A new stress test that verifies it is safe to concurrently get consumers for the same partition from the consumer pool.

tdas committed:
Fixed

tdas committed:
Removed println

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1387/
Test PASSed.

SparkQA commented:
**[Test build #88081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88081/testReport)** for PR 20767 at commit [`9e771b0`](https://github.com/apache/spark/commit/9e771b0125454fa8d72f00d19225a045ef635aac).

tdas commented:
jenkins retest this

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1388/
Test PASSed.

SparkQA commented:
**[Test build #88082 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88082/testReport)** for PR 20767 at commit [`9e771b0`](https://github.com/apache/spark/commit/9e771b0125454fa8d72f00d19225a045ef635aac).

SparkQA commented:
**[Test build #88081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88081/testReport)** for PR 20767 at commit [`9e771b0`](https://github.com/apache/spark/commit/9e771b0125454fa8d72f00d19225a045ef635aac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88081/
Test PASSed.

SparkQA commented:
**[Test build #88082 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88082/testReport)** for PR 20767 at commit [`9e771b0`](https://github.com/apache/spark/commit/9e771b0125454fa8d72f00d19225a045ef635aac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88082/
Test PASSed.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +401,65 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+    } else if (existingInternalConsumer == null) {
+      newNonCachedConsumer.internalConsumer.inuse = true
+      cache.put(key, newNonCachedConsumer.internalConsumer)
+      newNonCachedConsumer
```
We should return a CachedKafkaDataConsumer in this case. Right?

tdas replied:
oh yes. damn it. my bad.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +401,65 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+          consumer.inuse = false
+        }
       } else {
         logWarning(s"Attempting to release consumer that does not exist")
```
This is the case that a consumer may be evicted because of the max capacity. In this case, we should close the internal consumer.

tdas replied:
Aah. The warning was misleading. Will add comments to clarify that.

tdas replied:
This should not be the case. We do not remove any consumer from the cache while it is being used. So the scenario that you mentioned should not happen.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +401,65 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
-        cache.put(key, new CachedKafkaConsumer(topicPartition, kafkaParams))
+      // If this is reattempt at running the task, then invalidate cache and start with
+      // a new consumer
+      if (existingInternalConsumer != null) {
```
This logic here seems wrong. I think it should be something like this?
```
      if (existingInternalConsumer != null) {
        if (existingInternalConsumer.inuse) {
          existingInternalConsumer.markedForClose = true
          newNonCachedConsumer
        } else {
          existingInternalConsumer.close()
          cache.put(key, newNonCachedConsumer.internalConsumer)
          new CachedKafkaDataConsumer(newNonCachedConsumer.internalConsumer)
        }
      } else {
        cache.put(key, newNonCachedConsumer.internalConsumer)
        new CachedKafkaDataConsumer(newNonCachedConsumer.internalConsumer)
      }
```

tdas replied:
This is indeed better. What I was doing was always deferring to a later point. But that would lead to it being used one more time before being closed.

zsxwing reviewed (changes requested):
Left some comments

zsxwing commented on external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDataConsumerSuite.scala:
```diff
@@ -0,0 +1,111 @@
+    val numThreads = 50
+    val numConsumerUsages = 500
+
+    val threadpool = Executors.newFixedThreadPool(numThreads)
```
nit: threadpool should be shut down

tdas committed:
Fixed bugs

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1414/
Test PASSed.

SparkQA commented:
**[Test build #88109 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88109/testReport)** for PR 20767 at commit [`0a838c1`](https://github.com/apache/spark/commit/0a838c190ab75de2fc1ba9bd37ff8df37287d185).

SparkQA commented:
**[Test build #88109 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88109/testReport)** for PR 20767 at commit [`0a838c1`](https://github.com/apache/spark/commit/0a838c190ab75de2fc1ba9bd37ff8df37287d185).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88109/
Test PASSed.

brkyvz commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,99 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
   /**
    * Get a cached consumer for groupId, assigned to topic and partition.
    * If matching consumer doesn't already exist, will be created using kafkaParams.
+   * This will make a best effort attempt to
```
I would love to see the rest of this sentence. Such a cliffhanger!

brkyvz reviewed (approved):
Love this. Left one comment.

tdas committed:
Updated docs

SparkQA commented:
**[Test build #88140 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88140/testReport)** for PR 20767 at commit [`37a9225`](https://github.com/apache/spark/commit/37a92250e013c06ba10cf3ba3a720a8a347dd573).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1439/
Test PASSed.

SparkQA commented:
**[Test build #88140 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88140/testReport)** for PR 20767 at commit [`37a9225`](https://github.com/apache/spark/commit/37a92250e013c06ba10cf3ba3a720a8a347dd573).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88140/
Test PASSed.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -27,30 +27,73 @@ import org.apache.kafka.common.TopicPartition
   private val groupId = kafkaParams.get(ConsumerConfig.GROUP_ID_CONFIG).asInstanceOf[String]
 
-  private var consumer = createConsumer
+  @volatile private var consumer = createConsumer
```
I think these `@volatile`s are not necessary. I'm okey with them though.

tdas replied:
yeah, i just added them to be safer. one less thing to worry about.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+
+    } else {
+      // If consumer is already cached and is currently not in use, then return that consumer
+      CachedKafkaDataConsumer(existingInternalConsumer)
```
we should set `existingInternalConsumer.isuse = true`

tdas replied:
I wonder why this was not caught in the stress test.

tdas replied:
i will run a longer stress test locally just to be sure.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
-      val consumer = cache.get(key)
-      consumer.inuse = true
-      consumer
+    } else if (!useCache) {
```
this `if` should be moved before the above `if`.

tdas replied:
why? I am saying that we should NOT reuse consumers for ANY task retries, independent of the desire to use the cache or not.

zsxwing replied:
When `useCache` is `false`, i would expect `newInternalConsumer` should never be put into the cache. The above `if` may put `newInternalConsumer` into the cache. If we put a consumer used by a continuous processing query into the cache and assume it never ends, it will prevent other micro batch queries from putting a consumer reading the same topic partition into the cache.

tdas replied:
Technically that wont happens because the continuous query and the batch query will have different groupids. But I agree that if `useCache` is `false`, then we should not put it in the cache in any way. In fact, we can simplify the task retry case further by never putting the new one in the cache, only invalidate the existing cached one.

zsxwing commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+      if (cachedIntConsumer != null) {
+        if (cachedIntConsumer.eq(intConsumer)) {
+          // The released consumer is indeed the cached one.
+          cache.remove(key)
```
We should remove it only when it's closed.

tedyu commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
-      if (removedConsumer != null) {
-        removedConsumer.close()
+      // If it has been marked for close, then do it any way
+      if (intConsumer.inuse && intConsumer.markedForClose) intConsumer.close()
```
Is it possible we have the following condition - should intConsumer.close() be called ?

!intConsumer.inuse && intConsumer.markedForClose

tdas replied:
I rewrote the logic. Hopefully, it's simpler to reason about it now.

tedyu commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+        }
+      } else {
+        // Consumer is not cached, put the new one in the cache
+        cache.put(key, newInternalConsumer)
```
Should ```newInternalConsumer.inuse = true``` be called ?

tdas replied:
yes. correct. thanks!

tdas commented:
@koeninger good question Cody! I think we should fix this limitation eventually. The only reason I am not doing that in this PR is to keep the changes minimum for backporting to 2.3.x. Eventually, we should not do such cache management, and rather use something like [Apache Common Pool](https://commons.apache.org/proper/commons-pool/index.html).

tdas committed:
Simplified logic

tdas commented:
@tedyu @zsxwing thank you very much for catching the bugs. I have simplified the logic quite a bit. Note that I removed the invariant that I had introduced earlier. Additionally, I locally ran the stress test with 100 threads and 10000 read attempts, which ran for 2 mins. It passed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1545/
Test PASSed.

SparkQA commented:
**[Test build #88285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88285/testReport)** for PR 20767 at commit [`5363ea8`](https://github.com/apache/spark/commit/5363ea8c4a22340fda257658aba122839fe5a15a).

SparkQA commented:
**[Test build #88285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88285/testReport)** for PR 20767 at commit [`5363ea8`](https://github.com/apache/spark/commit/5363ea8c4a22340fda257658aba122839fe5a15a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88285/
Test PASSed.

tedyu commented on external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala:
```diff
@@ -467,44 +435,58 @@ private[kafka010] object KafkaDataConsumer extends Logging {
-    } else if (existingInternalConsumer.inuse) {
+    } else if (existingInternalConsumer.inUse) {
       // If consumer is already cached but is currently in use, then return a new consumer
       NonCachedKafkaDataConsumer(newInternalConsumer)
```
Maybe keep an internal counter for how many times the non cached consumer is created.
This would give us information on how effective the cache is

tdas commented:
The idea is good. But how do you propose exposing that information?
Periodic print out in the log?

From a different angle, I would rather not do feature creep in this PR that
is intended to be backported to 2.3.

On Mar 15, 2018 7:31 PM, "tedyu" <notifications@github.com> wrote:

> *@tedyu* commented on this pull request.
> ------------------------------
>
> In external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/
> KafkaDataConsumer.scala
> <https://github.com/apache/spark/pull/20767#discussion_r174984237>:
>
> >        CachedKafkaDataConsumer(newInternalConsumer)
>
> -    } else if (existingInternalConsumer.inuse) {
> +    } else if (existingInternalConsumer.inUse) {
>        // If consumer is already cached but is currently in use, then return a new consumer
>        NonCachedKafkaDataConsumer(newInternalConsumer)
>
> Maybe keep an internal counter for how many times the non cached consumer
> is created.
> This would give us information on how effective the cache is
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/20767#pullrequestreview-104439601>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAoerMcXNmKmobW4ws25hx3OvcER-1Ptks5teyPogaJpZM4SiC1I>
> .
>

zsxwing commented:
@tdas this is much simpler!!! LGTM. Merging to master.

tedyu commented:
@tdas 
Do you think a follow on JIRA can be logged for adding metrics for the cache operations ?

Thanks

zsxwing commented:
@tedyu that's a good idea. Could you create a ticket? Thanks!

tdas commented:
@tedyu @zsxwing My thoughts on this is that we should consider migrating to something like Apache Common Pool (assuming it does not require additional maven libraries), which might be less maintenance load. It could be that it already has metrics, etc. that we can leverage.

tedyu commented:
I did a quick search for 'apache commons pool metrics' which didn't show up directly related links.

tdas commented:
@tedyu It was indeed hard to find :) But apache commons pool does expose metrics on idle/active counts. See https://commons.apache.org/proper/commons-pool/apidocs/org/apache/commons/pool2/impl/BaseGenericObjectPool.html

tedyu commented:
Interesting.

https://commons.apache.org/proper/commons-pool/apidocs/org/apache/commons/pool2/impl/BaseGenericObjectPool.html#getBorrowedCount()

tdas commented:
Just to be clear, I am not saying that we *have to* move to this pool stuff. I am just saying that if we want to make this more robust, then we should try to use existing tools (after careful evaluation), rather than trying to reinvent the wheel.

gaborgsomogyi commented:
@tdas @zsxwing @koeninger @tedyu do you think it makes sense to make similar step in the DStream area like this and then later follow with the mentioned Apache Common Pool?
